torch/csrc/autograd/record_function.h

```diff
@@ -26,32 +28,70 @@ struct TORCH_API StringView {
   const char* str_ptr_;
 };
 
-using GetPackedInputsCallback = std::function<std::vector<c10::IValue>()>;
-
 struct TORCH_API RecordFunction {
-  explicit RecordFunction(Function* fn, GetPackedInputsCallback cb = nullptr);
-
-  explicit RecordFunction(
-      std::string name,
-      int64_t current_sequence_nr = -1,
-      GetPackedInputsCallback cb = nullptr);
+  // Default constructor is used with before function called afterwards
+  RecordFunction();
+
+  // before function initializes RecordFunction members and calls
+  // start callbacks;
+  // before function can be used only when RecordFunction is
+  // constructed with default constructor RecordFunction::RecordFunction()
+  void before(const char* name, int64_t sequence_nr = -1);
+  void before(std::string name, int64_t sequence_nr = -1);
+  void before(Function* fn, int64_t sequence_nr = -1);
+
+  template<typename F, typename... Args>
+  void before(
+      F fn,
+      int64_t current_sequence_nr,
+      const Args&... args) {
```

Comment:
nit: prefer `Args&&... args`. This will decay to `T` if it's a temporary value and to a reference otherwise